File: 2017/day15/part2.scala
class Generator(var current: Long, val factor: Long, val mod: Long) {
  def next: Long = {
    current = (current * factor) % 2147483647
    if (current % mod == 0) current
    else next
  }
}

def calculate(a: Generator, b: Generator): Long = {
  def loop(total: Long, pairs: Long): Long = {
    if (pairs == 5000000) return total
    if (a.next % 65536 == b.next % 65536) loop(total + 1, pairs + 1)
    else loop(total, pairs + 1)
  }
  loop(0, 0)
}

val a = new Generator(65, 16807, 4)
val b = new Generator(8921, 48271, 8)
println(calculate(a, b))

File: 2017/day15/part2.py
def gen(n, factor, mod):
    while True:
        n = (n* factor) % 2147483647
        if n % mod == 0:
            yield n

def b16_match(a, b):
    return a % 65536 == b % 65536 # 2**16

a = gen(int(raw_input("Generator A: [65] ") or 65), 16807, 4)
b = gen(int(raw_input("Generator B: [8921] ") or 8921), 48271, 8)
total = 0
for i in xrange(0, 5000000):
    if b16_match(a.next(), b.next()):
        total +=1

print "Result is: {}".format(total)
